pkg/crypto/keys/publickey.go

```diff
@@ -306,3 +316,83 @@ func (p *PublicKey) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+// KeyRecover recovers public key from the given signature (r, s) on the given message hash using given elliptic curve.
+// Algorithm source: SEC 1 Ver 2.0, section 4.1.6, pages 47-48 (https://www.secg.org/sec1-v2.pdf).
+// Flag isEven denotes Y's least significant bit in decompression algorithm.
+func KeyRecover(curve elliptic.Curve, r, s *big.Int, messageHash []byte, isEven bool) (PublicKey, error) {
+	var (
+		res PublicKey
+		err error
+	)
+	if r.Cmp(big.NewInt(1)) == -1 || s.Cmp(big.NewInt(1)) == -1 {
+		return res, errors.New("invalid signature")
+	}
+	params := curve.Params()
+	// calculate h = (Q + 1 + 2 * Sqrt(Q)) / N
+	// actually, we can skip this step for secp256k1 and secp256r1 as far as we know cofactor of these curves (h=1)
+	// (see section 2.4 of http://www.secg.org/sec2-v2.pdf)
+	num := new(big.Int).Add(new(big.Int).Add(params.P, big.NewInt(1)), new(big.Int).Mul(big.NewInt(2), new(big.Int).Sqrt(params.P)))
+	h := new(big.Int).Div(num, params.N)
```

Comment:
So if `we can skip this step for secp256k1 and secp256r1` maybe we can hardcode it (leaving commented out `h` calculations just in case they're needed in the future)? A little optimization is always good to have and it costs nothing from what I see.